include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+   * outstanding references to the data stored on this cell. That is to say,
+   * that the only references to the stored data are that made by this class.
+   */
+  bool erase(const CellIteratorType &cell);
+
+  /**
+   * Clear all the data stored in this object.
+   */
+  void clear();
+
+  /**
+   * Get a vector of the data located at @p cell .
+   * A possible additional typename @p T is the class to which the base class
+   * DataType could be cast. Since @p DataType is stored as shared pointers,
+   * there is minimal overhead in returning a vector by value instead of by reference.
+   * This allows flexibility if class @p T is not the same as @p DataType on a
+   * cell-by-cell basis.
+   */
+  template<typename T=DataType>
+  CellVector<T> get_data(const CellIteratorType &cell);
+
+  /**
+   * Get a vector of constant pointers to data located at @p cell .
+   * A possible additional typename @p T is the class to which the base class
+   * DataType could be cast. Since @p DataType is stored as shared pointers,
+   * there is minimal overhead in returning a vector by value instead of by reference.
+   * This allows flexibility if class @p T is not the same as @p DataType on a
+   * cell-by-cell basis.
+   */
+  template<typename T=DataType>
+  CellVector<const T> get_data(const CellIteratorType &cell) const;
+
+private:
+  /**
+   * A map to store a vector of data on a cell.
+   */
+  std::map<CellIteratorType, CellVector<DataType> > map;
+
+};
+
+
+/**
+ * An abstract class which specifies requirements for data on
+ * a single quadrature point to be transferable during refinement.
+ * One possible implementation is via the restriction/prolongation matrices
+ * as implemented in ContinuousQuadratureDataTransfer class.
+ *
+ * @author Denis Davydov, Jean-Paul Pelteret, 2016
+ */
+class TransferableQuadraturePointData
+{
+public:
+  /**
+   * Default constructor.
+   */
+  TransferableQuadraturePointData();
+
+  /**
+   * Default virtual destructor.
+   */
+  virtual ~TransferableQuadraturePointData();
+
+  /**
+   * Returns the total number of values which will be
+   * packed/unpacked from the user's DataType class. Consequently it is also
+   * the size of the vectors in pack_values() and unpack_values() .
+   */
+  virtual unsigned int number_of_values() const = 0;
+
+  /**
+   * Pack all data stored in this class into a vector @p values .
+   * This vector will contain all scalar and/or Tensorial data local to each
+   * quadrature point.
+   *
+   * @p values will be resized to number_of_values() .
+   */
+  virtual void pack_values(std::vector<double> &values) const = 0;
+
+  /**
+   * The opposite of the above, namely
+   * unpack a vector @p values into the data stored in this class.
+   *
+   * @p values will be resized to number_of_values() .
+   */
+  virtual void unpack_values(const std::vector<double> &values) = 0;
+};
+
+
+#ifdef DEAL_II_WITH_P4EST
+namespace distributed
```

Comment:
Hm, I'm not in favor of this. In the doxygen documentation, it will show up in separate namespaces from the triangulation classes then. I would wrap everything into namespace `parallel`.